pkg/commands/analyse_queries.go

```diff
@@ -0,0 +1,95 @@
+package commands
+
+import (
+	"bufio"
+	"io"
+	"sort"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+type MetricUsage struct {
+	LabelsUsed []string
+}
+
+func processQueries(r io.Reader) (map[string]MetricUsage, error) {
+	metrics := map[string]MetricUsage{}
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		if err := processQuery(scanner.Text(), metrics); err != nil {
+			return nil, err
+		}
+	}
+
+	return metrics, scanner.Err()
+}
+
+func processQuery(query string, metrics map[string]MetricUsage) error {
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return err
+	}
+
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		vs, ok := node.(*parser.VectorSelector)
+		if !ok {
+			return nil
+		}
+
+		metricName, ok := getName(vs.LabelMatchers)
+		if !ok {
+			return nil
+		}
+
+		usedLabels := metrics[metricName]
+
+		// Add any label names from the selectors to the list of used labels.
+		for _, matcher := range vs.LabelMatchers {
+			if matcher.Name == labels.MetricName {
+				continue
+			}
+			setInsert(matcher.Name, &usedLabels.LabelsUsed)
+		}
+
+		// Find any aggregations in the path and add grouping labels.
+		for _, node := range path {
+			ae, ok := node.(*parser.AggregateExpr)
+			if !ok {
+				continue
+			}
+
+			for _, label := range ae.Grouping {
+				setInsert(label, &usedLabels.LabelsUsed)
```

Comment:
why not just use a temporary `map` for `usedLabels.LabelsUsed`?

Reply:
No particular reason, I guess I was expecting this list to be short and this to be cheaper than all the allocs associated with maps.